### PR TITLE
Add support for Keyed Snowflake Sink

### PR DESCRIPTION
**Pull request by saiharshavellanki**
- Added tests and updated documentation

Screenshot 2024-10-13 at 8 00 47 AM

Screenshot 2024-10-13 at 8 06 44 AM

Screenshot 2024-10-13 at 8 03 24 AM

----

> [!IMPORTANT]
> Add support for Snowflake as a data sink, including updates to connectors, tests, and documentation.
> 
>   - **Behavior**:
>     - Add support for Snowflake as a data sink in `connectors.py`.
>     - Update `sink()` function to handle Snowflake-specific logic, including validation for CDC and cursor settings.
>     - Ensure only "incremental" style is supported for Snowflake sinks.
>   - **Connectors**:
>     - Modify `TableConnector` to support Snowflake sinks by allowing `cursor` to be optional.
>     - Add `_snowflake_conn_to_sink_proto()` in `to_proto.py` for Snowflake sink serialization.
>   - **Tests**:
>     - Add tests in `test_connectors.py` and `test_invalid_connectors.py` to validate Snowflake sink behavior.
>   - **Documentation**:
>     - Update `api-reference` docs to include Snowflake sink details.
>     - Adjust `api.yml` to categorize Snowflake under sink connectors.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=fennel-ai%2Fclient&utm_source=github&utm_medium=referral)<sup> for 1e61c38e64e8027a07176d320a1c803579ca14f3. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Review by nikhilgarg28 — APPROVED** [review]

File: docs/pages/concepts/sink.md [hunk 1]
```diff
@@ -24,8 +24,6 @@ the CDC data needs to be written out in the debezium format.
...
 
-As of right now, Fennel only supports Kafka sinks and writes data in the debezium
-format. Given the ubiquity of debezium connectors, you should be able to further 
-pipe this debezium data from Kafka to your data store of choice.
-
-More data stores and cdc strategies will be supported in the future updates.
\ No newline at end of file
+Fennel ships with data sinks to a couple of [common datastores](/api-reference/sink_connectors) 
+so that you can 'sink' from your Fennel datasets to your external datasets.
+We're continuously expanding support to include additional datastores.
```
Comment (nikhilgarg28) [inline docs/pages/concepts/sink.md:29]: nit: please reword this line a bit - guideline is to never use 1st person (we) in docs. Instead, please change it to `Sinks to many other common data stores will be added soon`

File: docs/pages/api-reference/sink_connectors/kafka.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+<Expandable title="security_protocol" type='"PLAINTEXT" | "SASL_PLAINTEXT" | "SASL_SSL"'>
+Protocol used to communicate with the brokers. 
+</Expandable>
+
+<Expandable title="sasl_mechanism" type='"PLAIN" | "SCRAM-SHA-256" | "SCRAM-SHA-512" | "GSSAPI"'>
+SASL mechanism to use for authentication. 
+</Expandable>
+
+<Expandable title="sasl_plain_username" type="Optional[str]">
```
Comment (nikhilgarg28) [inline docs/pages/api-reference/sink_connectors/kafka.md:33]: Can this be given as secret that Hemanth built? If so, maybe update the type signature?

File: docs/pages/api-reference/sink_connectors/kafka.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+SASL username.
+</Expandable>
+
+<Expandable title="sasl_plain_password" type="Optional[str]">
```
Comment (nikhilgarg28) [inline docs/pages/api-reference/sink_connectors/kafka.md:37]: Same here.

File: docs/pages/api-reference/sink_connectors/s3.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+---
+### S3
+Data connector to sink data to S3.
+
+#### Account Parameters
+<Expandable title="name" type="str">
+A name to identify the sink. The name should be unique across all Fennel connectors.
+</Expandable>
+
+<Expandable title="aws_access_key_id" type="Optional[str]" defaultVal="None">
```
Comment (nikhilgarg28) [inline docs/pages/api-reference/sink_connectors/s3.md:14]: Can this take secret? Could you do the same check for all other fields added for sinks?

**Review by nikhilgarg28 — COMMENTED** [review]
Requesting couple of nits, rest is great.

File: docs/pages/api-reference/sink_connectors/kafka.md (hunk 2 above)
Comment (saiharshavellanki) [reply docs/pages/api-reference/sink_connectors/kafka.md:33]: @hemanthk269 would be updating the documentation with Secrets entirely soon

File: docs/pages/api-reference/sink_connectors/kafka.md (hunk 3 above)
Comment (saiharshavellanki) [reply docs/pages/api-reference/sink_connectors/kafka.md:37]: @hemanthk269 would be updating the documentation with Secrets entirely soon

File: docs/pages/api-reference/sink_connectors/s3.md (hunk 4 above)
Comment (saiharshavellanki) [reply docs/pages/api-reference/sink_connectors/s3.md:14]: @hemanthk269 would be updating the documentation with Secrets entirely soon

**Pull request merged**